pyqir/src/values.rs

```diff
@@ -75,37 +76,110 @@ impl Value {
 }
 
 impl Value {
+    pub(crate) unsafe fn new(owner: Owner, value: AnyValue) -> PyClassInitializer<Self> {
+        let value = transmute::<AnyValue<'_>, AnyValue<'static>>(value);
+        PyClassInitializer::from(Self { value, owner })
+    }
+
     pub(crate) unsafe fn from_any<'ctx>(
         py: Python,
-        context: Py<Context>,
+        owner: Owner,
         value: impl Into<AnyValue<'ctx>>,
     ) -> PyResult<PyObject> {
         let value = transmute::<AnyValue<'_>, AnyValue<'static>>(value.into());
         #[allow(clippy::same_functions_in_if_condition)]
         if let Ok(inst) = value.try_into() {
-            Instruction::from_inst(py, context, inst)
+            Instruction::from_inst(py, owner, inst)
         } else if let Ok(block) = value.try_into() {
-            let base = PyClassInitializer::from(Self { value, context });
+            let base = PyClassInitializer::from(Self { value, owner });
             let block = base.add_subclass(BasicBlock(block));
             Ok(Py::new(py, block)?.to_object(py))
         } else if value.is_const() {
-            Constant::from_any(py, context, value)
+            Constant::from_any(py, owner, value)
         } else {
-            Ok(Py::new(py, Self { value, context })?.to_object(py))
+            Ok(Py::new(py, Self { value, owner })?.to_object(py))
         }
     }
 
-    pub(crate) unsafe fn init(context: Py<Context>, value: AnyValue) -> PyClassInitializer<Self> {
-        let value = transmute::<AnyValue<'_>, AnyValue<'static>>(value);
-        PyClassInitializer::from(Self { value, context })
-    }
-
     pub(crate) unsafe fn get(&self) -> AnyValue<'static> {
         self.value
     }
 
-    pub(crate) fn context(&self) -> &Py<Context> {
-        &self.context
+    pub(crate) fn owner(&self) -> &Owner {
+        &self.owner
+    }
+}
+
+/// To store Inkwell values in Python classes, we transmute the lifetime to `'static`. You need to
+/// be careful when using Inkwell types with unsafely extended lifetimes. Follow these rules:
+///
+/// 1. When storing in a data type, always include an `Owner` field containing the owning module, if
+///    there is one, or the context otherwise.
+/// 2. Before passing an LLVM object to an Inkwell function, call `Owner::merge` to ensure that the
+///    owners of all of the objects are compatible.
+pub(crate) enum Owner {
+    Context(Py<Context>),
+    Module(Py<Module>),
+}
+
+impl Owner {
+    pub(crate) fn context(&self, py: Python) -> Py<Context> {
+        match self {
+            Self::Context(context) => context.clone_ref(py),
+            Self::Module(module) => module.borrow(py).context().clone_ref(py),
+        }
+    }
+
+    pub(crate) fn merge(
+        py: Python,
+        owners: impl IntoIterator<Item = impl Borrow<Self>>,
+    ) -> PyResult<Self> {
+        owners
```

Comment:
I think we need a code comment on this merge impl

Reply:
I added a doc comment.